### PR TITLE
Add support to fetch from specfiles

### PR DESCRIPTION
**Pull request by obreitwi**
This change mirrors the `install` command's ability to take spec-input
via specfiles.

Same as for `install` they can be supplied via `--file`.

### CODE REVIEW

File: lib/spack/spack/cmd/fetch.py [hunk 1]
```diff
@@ -39,9 +43,22 @@ def setup_parser(subparser):
...
     if args.specs:
-        specs = spack.cmd.parse_specs(args.specs, concretize=True)
-    else:
+        specs += spack.cmd.parse_specs(args.specs, concretize=True)
+    if args.specfiles:
+        for specfile in args.specfiles:
+            with open(specfile, 'r') as f:
+                s = spack.spec.Spec.from_yaml(f)
+
+            if s.concretized().dag_hash() != s.dag_hash():
```
Comment (scheibelp) [inline lib/spack/spack/cmd/fetch.py:54]: You should be able to just say `if not s.concrete` here

**Comment by scheibelp** [comment]
I'll be able to respond to this by tomorrow (June 15th)

(I need to check a couple things WRT the internals)

**Review by scheibelp — CHANGES_REQUESTED** [review]
I have a suggestion for simplifying the logic.

Also since this is used by `spack install` as well as `spack fetch`, I think that this argument should be promoted to `cmd/common/arguments.py`. The problem with that is that `--file` is a bit too generic a name for an argument that refers to Spec YAML files, so I also think the name should change (`specfile` sounds good to me but could you check if that's used elsewhere?).

However, `-f` for `spack install` is used by some of our internal CI logic and possibly by other folks, so we should leave `spack install` be for now (other than perhaps adding a note there that we should deprecate the `-f` argument there in the next major release and switch to the common argument).

So in short:

* Change option name in `spack fetch` to `--specfile` (assuming that's not taken)
* Add this to `cmd/common/arguments.py`
* Add a TODO to `spack install` to deprecate `-f`

Does that sound workable?

**Comment by obreitwi** [comment]
Thank you for taking the time to review the PR!

I personally dislike the idea of having the same functionality available as two different arguments (`--specfile` for `fetch` and still `--file` for `install`).
Preferably, we should attempt one of the following:

1. Add `--specfile` to `install`/`fetch`, but additionally keep deprecated `-f/--file` for `install` with a warning.
2. Add `--file` as is to `fetch` to keep consistency with `install`.

I would prefer if users did not have to remember different names for the same functionality in different commands.

Which solution do you prefer? I tend to like the first more.

**Comment by scheibelp** [comment]
> 1. Add --specfile to install/fetch, but additionally keep deprecated -f/--file for install with a warning.

This sounds good. I don't think adding `-f` to `fetch` is necessary, but I think adding `--specfile` to `install` would be useful (and it can be processed in the same way as `-f`).

**Comment by alecbcs** [comment]
@obreitwi did you get a chance to take a look at the review feedback. Is this PR still something you'd like to pursue?

**Comment by obreitwi** [comment]
@alecbcs I will unfortunately not have time to pursue this PR further. Feel free to close this PR or take it over…
